courses/advanced-ada/advanced_topics.rst

```diff
@@ -1,5 +1,294 @@
+        - [ ] ``select...then`` statement
+    - [ ] Real-time programming (*material reuse*)
+    - [ ] High-Performance (*material reuse*)
+    - [ ] Task IDs and attributes
+    - [ ] Task termination
+    - [ ] Tasking and exceptions
+    - [ ] Ravenscar profile
+
+- [ ] File I/O
+
+    - [ ] Efficient Stream I/O for Array Types (*material reuse*)
+    - [ ] Container streaming
+
+- [ ] Numerics
+
+    - [ ] Modular types
+
+- [ ] Design Patterns
+
+    - [ ] Factory Functions (*material reuse*)
+    - [ ] Scope Locks Idiom (*material reuse*)
+    - [ ] Visitor Pattern (*material reuse*)
+    - [ ] Overridable Class Attributes (*material reuse*)
+
+    - Based on
+      `list of design patterns <https://en.wikipedia.org/wiki/Software_design_pattern>`_
+      and `list for C++ <https://en.wikibooks.org/wiki/C%2B%2B_Programming/Code/Design_Patterns>`_
+
+        - **Note**: the goal is to identify equivalent patterns for Ada
+
+        - Creational Patterns
+
+            - [ ] Builder
+            - [ ] Factory / Abstract Factory
+            - [ ] Lazy initialization
+            - [ ] Multiton
+            - [ ] Object pool
+            - [ ] Dependency Injection
+            - [ ] Prototype
+            - [ ] Resource acquisition is initialization (RAII)
+            - [ ] Singleton
+
+        - Structural Patterns
+
+            - [ ] Adapter or Wrapper
+            - [ ] Bridge
+            - [ ] Composite
+            - [ ] Decorator
+            - [ ] Extension object
+            - [ ] Façade
+            - [ ] Flyweight
+            - [ ] Front controller
+            - [ ] Marker
+            - [ ] Module (probably not worth mentioning)
+            - [ ] Proxy
+            - [ ] Twin
+            - [ ] Mixin
+            - [ ] Interface-based Programming (IBP)
+
+        - Behavioral Patterns
+
+            - [ ] Blackboard
+            - [ ] Chain of Responsibility
+            - [ ] Command
+            - [ ] Interpreter
+            - [ ] Iterator
+            - [ ] Mediator
+            - [ ] Memento
+            - [ ] Null object
+            - [ ] Observer or Publish/subscribe
+            - [ ] Servant
+            - [ ] Specification
+            - [ ] State
+            - [ ] Strategy
+            - [ ] Template Method
+            - [ ] Visitor
+            - [ ] Model-View-Controller (MVC)
+
+        - Concurrency patterns
+
+            - [ ] Active Object
+            - [ ] Balking
+            - [ ] Binding properties
+            - [ ] Compute kernel
+            - [ ] Double-checked locking
+            - [ ] Event-based asynchronous
+            - [ ] Guarded suspension
+            - [ ] Messaging design pattern (MDP)
+            - [ ] Monitor object
+            - [ ] Reactor
+            - [ ] Read-write lock
+            - [ ] Scheduler
+            - [ ] Thread pool
+            - [ ] Thread-specific storage
+
+    - [ ] Ada-specific design patterns
+
+        - Signature package
```

Comment:
This is not really Ada specific. It maps strongly to Haskell's [type classes](https://www.haskell.org/tutorial/classes.html) or [Rust traits](https://doc.rust-lang.org/rust-by-example/trait.html)

It maps even more strongly to [OCaml functors](https://v1.realworldocaml.org/v1/en/html/functors.html).

I don't know what it means :) But I would be careful at isolating particular patterns, or giving to much credit to the "official" lists.